src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java

```diff
@@ -36,494 +18,316 @@
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.SEVERE;
 
 /**
- * Listens to Jenkins Job build {@link Run} start and stop then ensure there's a suitable {@link Build} object in
- * OpenShift thats updated correctly with the current status, logsURL and metrics
+ * Listens to Jenkins Job build Step then ensure there's a suitable {@link Build} object in
+ * OpenShift that's updated correctly with the Job {@link Run} status
  */
 @Extension
-public class BuildSyncRunListener extends RunListener<Run> {
-  private static final Logger logger = Logger.getLogger(BuildSyncRunListener.class.getName());
+public class BuildSyncRunListener implements GraphListener {
+
   public static final String JENKINS_ROOT_URL_ENV_VAR = "JENKINS_ROOT_URL";
+  private static final Logger logger = Logger.getLogger(BuildSyncRunListener.class.getName());
 
-  private long pollPeriodMs = 1000 * 5;  // 5 seconds
-  private long delayPollPeriodMs = 1000; // 1 seconds
   private String namespace;
-  //Not used as we have disable this
-  //private static final long maxDelay = 30000;
-
-  private transient Set<Run> runsToPoll = new CopyOnWriteArraySet<>();
-
-  private transient AtomicBoolean timerStarted = new AtomicBoolean(false);
-  private Boolean kubernetesCluster;
+  private ExecutorService updateExecutor;
+  private OpenShiftClient openShiftClient;
 
   public BuildSyncRunListener() {
-    init();
+    openShiftClient = getOpenShiftClient();
+    namespace = OpenShiftUtils.getNamespaceOrUseDefault(namespace, openShiftClient);
+    updateExecutor = Executors.newSingleThreadExecutor();
   }
 
-  @DataBoundConstructor
-  public BuildSyncRunListener(long pollPeriodMs) {
-    this.pollPeriodMs = pollPeriodMs;
-    init();
-  }
-
-  /**
-   * Joins all the given strings, ignoring nulls so that they form a URL with / between the paths without a // if the
-   * previous path ends with / and the next path starts with / unless a path item is blank
-   *
-   * @param strings the sequence of strings to join
-   * @return the strings concatenated together with / while avoiding a double // between non blank strings.
-   */
-  public static String joinPaths(String... strings) {
-    StringBuilder sb = new StringBuilder();
-    for (int i = 0; i < strings.length; i++) {
-      sb.append(strings[i]);
-      if (i < strings.length - 1) {
-        sb.append("/");
-      }
-    }
-    String joined = sb.toString();
-
-    // And normalize it...
-    return joined
-      .replaceAll("/+", "/")
-      .replaceAll("/\\?", "?")
-      .replaceAll("/#", "#")
-      .replaceAll(":/", "://");
-  }
-
-  private void init() {
-    namespace = OpenShiftUtils.getNamespaceOrUseDefault(namespace, getOpenShiftClient());
+  public BuildSyncRunListener(OpenShiftClient ocClient, String ns) {
+    openShiftClient = ocClient;
+    namespace = OpenShiftUtils.getNamespaceOrUseDefault(ns, openShiftClient);
+    updateExecutor = Executors.newSingleThreadExecutor();
```

Comment:
> The executor needs to be static, where you create a pool of say 25, and where the pool size is configurable (a sys prop vs. an explicit config field is fine).

You mean using `newFixedThreadPool()` Or definate queue size?

> Certainly when you consider this is a per namespace thing per the ctor args and 

I'm sorry, I didn't get this point. Could you please elaborate it?

> we could see BuildSyncRunListener getting created many times over long running Jenkins instances.

Oh, thank you for sharing this observation. I wasn't aware of this case. So it's in general Jenkins behavior? My impression is `@Extension` class is a singleton and its get initialized only once throughout the lifecycle of Jenkins.

Reply:
General java behavior really, though Jenkins does not do anything to help.

But yeah, I got mixed up a bit by your Constructor with its params that I ignored the @Extension piece.   So 
a) in the fabric8 version of this plugin, you've got namespace set on this object, where presumably you have only 1 namesapce to deal with in your scenarios, right ?
b) yeah, the params in the constructor has to do with Jenkins rehydrating the object on restart

If I have that right, it is fine for you.  But if we are to pick this upstream, based on how we reconciled things, since we support multiple namespaces being monitored, we would want the executor pool globalized.

Still sorting through your the rest of your new commits, but let's sort out these details one way or the other.

Reply:
>  in the fabric8 version of this plugin, you've got namespace set on this object, where presumably you have only 1 namesapce to deal with in your scenarios, right ?

Yes

> If I have that right, it is fine for you. But if we are to pick this upstream, based on how we reconciled things, since we support multiple namespaces being monitored, we would want the executor pool globalized.

I see. On other note in this patch updater `update()` is taking namespace as the second parameter. So its no more fixed with a constructor. However, consumer of the class has to derive the right namespace.

Really appreciate for bringing this point.